pkg/tiller/release_update.go

```diff
@@ -137,6 +142,97 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
+	// find the last release with the given name
+	oldRelease, err := s.env.Releases.Last(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	newRelease, err := s.prepareRelease(&services.InstallReleaseRequest{
+		Chart:        req.Chart,
+		Values:       req.Values,
+		DryRun:       req.DryRun,
+		Name:         req.Name,
+		DisableHooks: req.DisableHooks,
+		Namespace:    oldRelease.Namespace,
+		ReuseName:    true,
+		Timeout:      req.Timeout,
+		Wait:         req.Wait,
+	})
+	res := &services.UpdateReleaseResponse{Release: newRelease}
+	if err != nil {
+		s.Log("failed update prepare step: %s", err)
+		// On dry run, append the manifest contents to a failed release. This is
+		// a stop-gap until we can revisit an error backchannel post-2.0.
+		if req.DryRun && strings.HasPrefix(err.Error(), "YAML parse error") {
+			err = fmt.Errorf("%s\n%s", err, newRelease.Manifest)
+		}
+		return res, err
+	}
+
+	// pre-upgrade hooks
+	if !req.DisableHooks {
+		if err := s.execHook(newRelease.Hooks, newRelease.Name, newRelease.Namespace, hooks.PreUpgrade, req.Timeout); err != nil {
+			return res, err
+		}
+	} else {
+		s.Log("update hooks disabled for %s", req.Name)
+	}
+
+	// From here on out, the release is considered to be in Status_DELETING or Status_DELETED
+	// state. There is no turning back.
+	oldRelease.Info.Status.Code = release.Status_DELETING
+	oldRelease.Info.Deleted = timeconv.Now()
+	oldRelease.Info.Description = "Deletion in progress (or silently failed)"
+	s.recordRelease(oldRelease, true)
+
+	// delete manifests from the old release
+	_, errs := s.ReleaseModule.Delete(oldRelease, nil, s.env)
+
+	oldRelease.Info.Status.Code = release.Status_DELETED
+	oldRelease.Info.Description = "Deletion complete"
+	s.recordRelease(oldRelease, true)
+
+	if len(errs) > 0 {
+		es := make([]string, 0, len(errs))
+		for _, e := range errs {
+			s.Log("error: %v", e)
+			es = append(es, e.Error())
+		}
+		return res, fmt.Errorf("Upgrade --force successfully deleted the previous release, but encountered %d error(s) and cannot continue: %s", len(es), strings.Join(es, "; "))
+	}
+
+	// update new release with next revision number so as to append to the old release's history
+	newRelease.Version = oldRelease.Version + 1
+	s.recordRelease(newRelease, false)
+	if err := s.ReleaseModule.Update(oldRelease, newRelease, req, s.env); err != nil {
+		msg := fmt.Sprintf("Upgrade %q failed: %s", newRelease.Name, err)
+		s.Log("warning: %s", msg)
+		newRelease.Info.Status.Code = release.Status_FAILED
+		newRelease.Info.Description = msg
+		s.recordRelease(newRelease, true)
+		return res, err
+	}
+
+	// post-upgrade hooks
+	if !req.DisableHooks {
+		if err := s.execHook(newRelease.Hooks, newRelease.Name, newRelease.Namespace, hooks.PostUpgrade, req.Timeout); err != nil {
+			return res, err
+		}
+	}
+
+	oldRelease.Info.Status.Code = release.Status_SUPERSEDED
```

Comment:
Hmm good point. I also think it might be good to just keep this in the FAILED state so others cannot roll back to this release, which SUPERSEDED or REPLACED would allow. The more I think about this, the more I would prefer to retain the existing FAILED state.

Related comment: https://github.com/kubernetes/helm/pull/3597#issuecomment-370050089

@adamreese any opinions on this?

Reply:
I think FAILED would also be a good idea